pinot-core/src/main/java/org/apache/pinot/core/segment/processing/reducer/RollupReducer.java

```diff
@@ -0,0 +1,172 @@
+      throws Exception {
+    LOGGER.info("Start reducing on partition: {}", _partitionId);
+    long reduceStartTimeMs = System.currentTimeMillis();
+
+    GenericRowFileReader fileReader = _fileManager.getFileReader();
+    int numRows = fileReader.getNumRows();
+    int numSortFields = fileReader.getNumSortFields();
+    LOGGER.info("Start sorting on numRows: {}, numSortFields: {}", numRows, numSortFields);
+    long sortStartTimeMs = System.currentTimeMillis();
+    GenericRowFileRecordReader recordReader = fileReader.getRecordReader();
+    LOGGER.info("Finish sorting in {}ms", System.currentTimeMillis() - sortStartTimeMs);
+
+    List<FieldSpec> fieldSpecs = _fileManager.getFieldSpecs();
+    boolean includeNullFields = _fileManager.isIncludeNullFields();
+    List<AggregatorContext> aggregatorContextList = new ArrayList<>();
+    for (FieldSpec fieldSpec : fieldSpecs) {
+      if (fieldSpec.getFieldType() == FieldType.METRIC) {
+        aggregatorContextList.add(new AggregatorContext(fieldSpec,
+            _aggregationTypes.getOrDefault(fieldSpec.getName(), DEFAULT_AGGREGATOR_TYPE)));
+      }
+    }
+
+    File partitionOutputDir = new File(_reducerOutputDir, _partitionId);
+    FileUtils.forceMkdir(partitionOutputDir);
+    LOGGER.info("Start creating rollup file under dir: {}", partitionOutputDir);
+    long rollupFileCreationStartTimeMs = System.currentTimeMillis();
+    GenericRowFileManager rollupFileManager =
+        new GenericRowFileManager(partitionOutputDir, fieldSpecs, includeNullFields, 0);
+    GenericRowFileWriter rollupFileWriter = rollupFileManager.getFileWriter();
+    GenericRow previousRow = new GenericRow();
+    recordReader.read(0, previousRow);
+    int previousRowId = 0;
+    GenericRow buffer = new GenericRow();
+    if (includeNullFields) {
+      for (int i = 1; i < numRows; i++) {
+        buffer.clear();
+        recordReader.read(i, buffer);
+        if (recordReader.compare(previousRowId, i) == 0) {
+          aggregateWithNullFields(previousRow, buffer, aggregatorContextList);
+        } else {
+          rollupFileWriter.write(previousRow);
+          previousRowId = i;
+          GenericRow temp = previousRow;
+          previousRow = buffer;
+          buffer = temp;
+        }
+      }
+    } else {
+      for (int i = 1; i < numRows; i++) {
+        buffer.clear();
+        recordReader.read(i, buffer);
+        if (recordReader.compare(previousRowId, i) == 0) {
+          aggregateWithoutNullFields(previousRow, buffer, aggregatorContextList);
+        } else {
+          rollupFileWriter.write(previousRow);
+          previousRowId = i;
+          GenericRow temp = previousRow;
+          previousRow = buffer;
+          buffer = temp;
+        }
+      }
+    }
+    rollupFileWriter.write(previousRow);
+    rollupFileManager.closeFileWriter();
+    LOGGER.info("Finish creating rollup file in {}ms", System.currentTimeMillis() - rollupFileCreationStartTimeMs);
+
+    _fileManager.cleanUp();
```

Comment:
how are we controlling the number of records per segment now?

Reply:
It is done during the segment creation phase, where we create the segment on a range of docs.
The reducer will always work on the whole partition. Since the sorting is done off-heap, it should not cause any memory issue.